kubetest/gke.go

```diff
@@ -487,15 +489,127 @@ func (g *gkeDeployer) TestSetup() error {
+			return fmt.Errorf("listing instances failed: %s", util.ExecError(err))
+		}
+		if len(output) == 0 {
+			continue
+		}
+		// Proxy instance found
+		fields := strings.Split(strings.TrimSpace(string(output)), "\t")
+		if len(fields) != 2 {
+			return fmt.Errorf("error parsing instances list output %q", output)
 		}
+		bastion, zone = fields[0], fields[1]
+		break
+	}
+	if bastion == "" {
+		return fmt.Errorf("proxy instance %q not found", g.sshProxyInstanceName)
+	}
+	log.Printf("Found proxy instance %q", bastion)
+
+	err := setKubeShhBastionEnv(g.project, zone, bastion)
+	if err == nil {
+		// Success, finishing
+		return nil
+	}
+	if !strings.Contains(err.Error(), "doesn't have an external IP address") {
+		return fmt.Errorf("setting KUBE_SSH_BASTION variable failed: %s", util.ExecError(err))
+	}
+
+	log.Printf("Trying to reserve external IP address for proxy instance")
+	g.bastionIPName = "bastion-ip-" + g.cluster
+
+	region, err := g.getRegion()
+	if err != nil {
+		return fmt.Errorf("error finding region for proxy instance address: %v", err)
+	}
+	if err := control.FinishRunning(exec.Command("gcloud", "compute", "addresses", "create", g.bastionIPName,
+		"--region="+region,
+		"--project="+g.project)); err != nil {
+		return fmt.Errorf("error creating address for proxy instance: %s", util.ExecError(err))
+	}
+	ipOut, err := exec.Command("gcloud", "compute", "addresses", "describe", g.bastionIPName,
+		"--format=value(address)",
+		"--region="+region,
+		"--project="+g.project).Output()
+	if err != nil {
+		return fmt.Errorf("error describing address for proxy instance: %s", util.ExecError(err))
+	}
+	bastionIPAddress := strings.TrimSpace(string(ipOut))
+	log.Printf("Reserved address: %s", bastionIPAddress)
+
+	log.Printf("Trying to attach external IP address to proxy instance")
+	if err := control.NoOutput(exec.Command("gcloud", "compute", "instances", "delete-access-config", bastion,
+		"--zone="+zone,
+		"--project="+g.project)); err != nil {
+		log.Printf("error cleaning up nat access config in proxy instance: %s", util.ExecError(err))
+	}
+	if err := control.FinishRunning(exec.Command("gcloud", "compute", "instances", "add-access-config", bastion,
+		"--address="+bastionIPAddress,
+		"--zone="+zone,
+		"--project="+g.project)); err != nil {
+		return fmt.Errorf("error attaching external IP address to proxy instance: %s", util.ExecError(err))
+	}
+	log.Printf("Attached address %s to %s", bastionIPAddress, bastion)
+
+	err = setKubeShhBastionEnv(g.project, zone, bastion)
+	if err != nil {
+		return fmt.Errorf("setting KUBE_SSH_BASTION variable failed: %s", util.ExecError(err))
+	}
+	return nil
+}
+
+func (g *gkeDeployer) cleanupBastion() error {
+	if g.bastionIPName == "" {
+		return nil
+	}
+	region, err := g.getRegion()
+	if err != nil {
+		return fmt.Errorf("error finding region for proxy instance address: %v", err)
+	}
+	if err := control.FinishRunning(exec.Command("gcloud", "compute", "addresses", "delete", g.bastionIPName,
```

Comment:
If the bastion VM had public IP then we won't have the address and the deletion will fail, right?

Reply:
I'm sorry, it was old version of the code (with static IPs). Somehow my next push landed in the wrong place.
But address deletion was guarded by new address name, not by bastion name.